### PR TITLE
Remove duplicate host routes from k8s dataclient

### PR DESCRIPTION
**Pull request by MustafaSaber**
Updates https://github.bus.zalan.do/pg9/issues/issues/33
Signed-off-by: Mustafa Abdelrahman <mustafa.abdelrahman@zalando.de>

### CODE REVIEW

File: dataclients/kubernetes/ingress.go [hunk 1]
```diff
@@ -399,7 +399,7 @@ func (ing *ingress) convert(state *clusterState, df defaultFilters, r *certregis
 			if err != nil {
 				return nil, err
 			}
-			if r != nil {
+			if r != nil && r.UniqueRoute(routes) {
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/ingress.go:402]: Both `LoadAll` and `LoadUpdate` use `mapRoutes(r) `:
https://github.com/zalando/skipper/blob/504a5e05e301bb1f19c7e6a1a07325d24d4a763f/dataclients/kubernetes/kube.go#L458-L484

Lets check if we can utilize it to detect duplicates at one place.

File: dataclients/kubernetes/kube.go [hunk 2]
```diff
@@ -351,13 +351,23 @@ func ParsePathMode(s string) (PathMode, error) {
...
 }
 
-func mapRoutes(r []*eskip.Route) map[string]*eskip.Route {
+func mapRoutes(r []*eskip.Route) (map[string]*eskip.Route, []*eskip.Route) {
+	var uniqueRoutes []*eskip.Route
 	m := make(map[string]*eskip.Route)
 	for _, ri := range r {
-		m[ri.Id] = ri
+		if _, found := m[ri.Id]; found {
+			if eskip.Eq(m[ri.Id], ri) {
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/kube.go:359]:
```go
if existing, ok := m[ri.Id]; ok {
  if eskip.Eq(existing, ri) {
  ...
  }
}
```

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-path.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  - host: www.example.org
+    http:
+      paths:
+      - path: "/test"
+        pathType: ImplementationSpecific
+        backend:
+          service:
+            name: bar
+            port:
+              name: baz
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-path.yaml:32]: This actually looks like a bug - they should not have the same id

File: go.mod [hunk 4]
```diff
@@ -19,7 +19,7 @@ require (
 	github.com/ghodss/yaml v1.0.0
 	github.com/go-redis/redis/v9 v9.0.0-beta.1
 	github.com/golang-jwt/jwt/v4 v4.2.0
-	github.com/google/go-cmp v0.5.8
+	github.com/google/go-cmp v0.5.9
```
Comment (AlexanderYastrebov) [inline go.mod:22]: Why do we have dependency changes?

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host.yaml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  - host: www.example.org
+    http:
+      paths:
+      - path: "/"
+        pathType: ImplementationSpecific
+        backend:
+          service:
+            name: bar
+            port:
+              name: baz
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host.yaml:32]: Lets also have a test case for the same host and the same path but different services

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-path.yaml (hunk 3 above)
Comment (AlexanderYastrebov) [reply dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-path.yaml:32]: The don't actually.

Comment (MustafaSaber) [reply dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-path.yaml:32]: It's to ensure duplicate host != duplicate routeID, will add another test case for different services.

File: dataclients/kubernetes/kube.go [hunk 6]
```diff
@@ -351,13 +351,23 @@ func ParsePathMode(s string) (PathMode, error) {
...
+	var uniqueRoutes []*eskip.Route
 	m := make(map[string]*eskip.Route)
 	for _, ri := range r {
-		m[ri.Id] = ri
+		if existingRoute, ok := m[ri.Id]; ok {
+			if eskip.Eq(existingRoute, ri) {
+				log.Warnf("Duplicate Ingress routeID found: %s", ri.Id)
+			} else {
+				log.Errorf("Duplicate Ingress routeID with different services found: %s\n%s is omitted.", ri.Id, ri.String())
+			}
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/kube.go:363]: It is not necessary coming from `Ingress` so lets remove it.

> with different services found

We don't know exactly what is the difference, lets just log the one we've omitted

File: dataclients/kubernetes/kube.go [hunk 7]
```diff
@@ -351,13 +351,23 @@ func ParsePathMode(s string) (PathMode, error) {
...
+func mapRoutes(r []*eskip.Route) (map[string]*eskip.Route, []*eskip.Route) {
+	var uniqueRoutes []*eskip.Route
 	m := make(map[string]*eskip.Route)
 	for _, ri := range r {
-		m[ri.Id] = ri
+		if _, found := m[ri.Id]; found {
+			if eskip.Eq(m[ri.Id], ri) {
+				log.Warnf("Duplicate Ingress route found: %s/%s", ri.Namespace, ri.Method)
+			} else {
+				log.Errorf("Duplicate Ingress route with different path found: %s/%s", ri.Namespace, ri.Method)
```
[the remark this answers is not in the document]
Comment (szuecs) [reply dataclients/kubernetes/kube.go]: @AlexanderYastrebov why is namespace deprecated?

File: dataclients/kubernetes/kube.go [hunk 8]
```diff
@@ -351,13 +351,23 @@ func ParsePathMode(s string) (PathMode, error) {
 	}
 }
 
-func mapRoutes(r []*eskip.Route) map[string]*eskip.Route {
+func mapRoutes(r []*eskip.Route) (map[string]*eskip.Route, []*eskip.Route) {
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/kube.go:354]: This would be called on each route poll cycle (every ~3s) so we risk spamming the logs.
Maybe we should ratelimit logging similar to `onPanicSometimes` in https://github.com/zalando/skipper/pull/2194

@szuecs WDYT?

File: go.mod [hunk 9]
```diff
@@ -34,7 +34,7 @@ require (
 	github.com/sirupsen/logrus v1.8.1
 	github.com/sony/gobreaker v0.5.0
 	github.com/stretchr/testify v1.8.0
-	github.com/szuecs/rate-limit-buffer v0.7.1
+	github.com/szuecs/rate-limit-buffer v0.7.2
```
Comment (szuecs) [inline go.mod:37]: this seems unrelated

File: go.sum [hunk 10]
```diff
@@ -816,8 +816,8 @@ github.com/stretchr/testify v1.8.0/go.mod h1:yNjHg4UonilssWZ8iaSj1OCr/vHnekPRkoO
 github.com/syndtr/gocapability v0.0.0-20170704070218-db04d3cc01c8/go.mod h1:hkRG7XYTFWNJGYcbNJQlaLq0fg1yr4J4t/NcTQtrfww=
 github.com/syndtr/gocapability v0.0.0-20180916011248-d98352740cb2/go.mod h1:hkRG7XYTFWNJGYcbNJQlaLq0fg1yr4J4t/NcTQtrfww=
 github.com/syndtr/gocapability v0.0.0-20200815063812-42c35b437635/go.mod h1:hkRG7XYTFWNJGYcbNJQlaLq0fg1yr4J4t/NcTQtrfww=
-github.com/szuecs/rate-limit-buffer v0.7.1 h1:kpVLwDvpCTFQi8uhiXQrhAKWzNUaEKhArFdjb4GQ8F4=
-github.com/szuecs/rate-limit-buffer v0.7.1/go.mod h1:BxqrsmnHsCnWcvbtdcaDLEBmjNEvRFU5LQ8edoZ9B0M=
+github.com/szuecs/rate-limit-buffer v0.7.2 h1:uJjt2hp3IgYxVed4doMnR8WC0Nifh8k8iOqrX1JKmqw=
+github.com/szuecs/rate-limit-buffer v0.7.2/go.mod h1:BxqrsmnHsCnWcvbtdcaDLEBmjNEvRFU5LQ8edoZ9B0M=
```
Comment (szuecs) [inline go.sum:820]: this seems unrelated

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+kind: Endpoints
+metadata:
+  labels:
+    application: myapp
+  namespace: foo
+  name: bar2
+subsets:
+- addresses:
+  - ip: 10.2.9.103
+  - ip: 10.2.9.104
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml:99]: Lets use different fake IPs for different endpoints

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+  - host: www.example.org
+    http:
+      paths:
+      - path: "/"
+        pathType: ImplementationSpecific
+        backend:
+          service:
+            name: bar
+            port:
+              name: baz
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml:32]: Service seems to be the same though - `bar`

File: go.mod (hunk 9 above)
Comment (AlexanderYastrebov) [inline go.mod:37]: I think this change is unrelated

File: dataclients/kubernetes/kube.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply dataclients/kubernetes/kube.go]: https://github.com/zalando/skipper/blob/76976749eb9348004c047b071165e83ef67214aa/eskip/eskip.go#L335-L340

File: dataclients/kubernetes/kube.go [hunk 13]
```diff
@@ -351,13 +351,23 @@ func ParsePathMode(s string) (PathMode, error) {
...
+	var uniqueRoutes []*eskip.Route
 	m := make(map[string]*eskip.Route)
 	for _, ri := range r {
-		m[ri.Id] = ri
+		if existingRoute, ok := m[ri.Id]; ok {
+			if eskip.Eq(existingRoute, ri) {
+				log.Warnf("Duplicate Ingress routeID found: %s", ri.Id)
+			} else {
+				log.Errorf("Duplicate Ingress routeID with different services found: %s\n%s is omitted.", ri.Id, ri.String())
```
[the remark this answers is not in the document]
Comment (MustafaSaber) [reply dataclients/kubernetes/kube.go:362]: According to the report we don't have and routegroups with duplicate hosts, also in code we return unique hosts fro RG https://github.com/zalando/skipper/blob/51a92b9740d209cd3fd6accd72495f23092e613b/dataclients/kubernetes/routegroup.go#L566

[the remark this answers is not in the document]
Comment (MustafaSaber) [reply dataclients/kubernetes/kube.go:362]: but removing `Ingress` from here would be better we can know the type from routeID anyway

File: dataclients/kubernetes/kube.go (hunk 8 above)
Comment (szuecs) [reply dataclients/kubernetes/kube.go:354]: yes maybe, but then we can also argue this for all the "route update" info/error logs

Comment (MustafaSaber) [reply dataclients/kubernetes/kube.go:354]: Maybe we should make it warning/alert that require fixing?

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml (hunk 11 above)
Comment (AlexanderYastrebov) [reply dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml:99]: Make it more realistic.
Different services normally have different selectors and thus select different pods.
Route backend endpoint list depends on the values of these IPs - with the same IPs routes could become equal while they should not.

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml (hunk 12 above)
Comment (AlexanderYastrebov) [reply dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml:32]: Should be fixed I guess?

File: dataclients/kubernetes/kube.go (hunk 8 above)
Comment (AlexanderYastrebov) [reply dataclients/kubernetes/kube.go:354]: Ok, lets keep it like that fro now.

> but then we can also argue this for all the "route update" info/error logs

lets do log ratelimiting for all logging in this module later then

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml (hunk 12 above)
Comment (MustafaSaber) [reply dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.yaml:32]: yes, will fix it. Just rebased to remove go.mod changes

File: dataclients/kubernetes/kube.go [hunk 14]
```diff
@@ -351,13 +351,23 @@ func ParsePathMode(s string) (PathMode, error) {
...
+func mapRoutes(r []*eskip.Route) (map[string]*eskip.Route, []*eskip.Route) {
+	var uniqueRoutes []*eskip.Route
 	m := make(map[string]*eskip.Route)
 	for _, ri := range r {
-		m[ri.Id] = ri
+		if existingRoute, ok := m[ri.Id]; ok {
+			if eskip.Eq(existingRoute, ri) {
+				log.Warnf("Duplicate routeID found: %s", ri.Id)
+			} else {
+				log.Errorf("Duplicate routeID with different services found: %s, %s is omitted.", ri.Id, ri.String())
```
Comment (szuecs) [inline dataclients/kubernetes/kube.go:362]: Error message should be consistent: "Failed to do ..."
Then I would think it might make sense to tell that two route X and Y (both route.Id are interesting) are found that differ in `cmp.Diff`.

File: dataclients/kubernetes/kube.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (szuecs) [reply dataclients/kubernetes/kube.go]: yes ok makes sense in eskip.

File: dataclients/kubernetes/kube.go [hunk 15]
```diff
@@ -351,13 +351,23 @@ func ParsePathMode(s string) (PathMode, error) {
...
+func mapRoutes(r []*eskip.Route) (map[string]*eskip.Route, []*eskip.Route) {
+	var uniqueRoutes []*eskip.Route
 	m := make(map[string]*eskip.Route)
 	for _, ri := range r {
-		m[ri.Id] = ri
+		if existingRoute, ok := m[ri.Id]; ok {
+			if eskip.Eq(existingRoute, ri) {
+				log.Warnf("Duplicate routeID found: %s", ri.Id)
+			} else {
+				log.Errorf("Failed to map route, duplicate ID with different body found, existing routeID: %s, new routeID: %s, new route body: %s", existingRoute.Id, ri.Id, ri.String())
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/kube.go:362]: existingRoute.Id equals ri.Id so there is no need to log it

File: dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.eskip [hunk 16]
```diff
@@ -0,0 +1,10 @@
+kube_foo__qux______: *
+	-> <roundRobin, "http://10.2.9.103:2134", "http://10.2.9.104:2134">;
+
+kube_foo__qux__www_example_org_____bar:
+	Host("^(www[.]example[.]org[.]?(:[0-9]+)?)$") && PathRegexp("^/")
+	-> <roundRobin, "http://10.2.9.103:8080", "http://10.2.9.104:8080">;
+
+kube_foo__qux__www_example_org_____bar2:
+	Host("^(www[.]example[.]org[.]?(:[0-9]+)?)$") && PathRegexp("^/")
+	-> <roundRobin, "http://10.2.9.105:8081", "http://10.2.9.106:8081">;
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/testdata/ingressV1/ingress-data/ing-with-duplicate-host-different-service.eskip:10]: So route id contains service name but not port name/number.
So if we have an ingress with two rules using the same service with two different ports we would imit the second route?...

File: dataclients/kubernetes/kube.go (hunk 6 above)
Comment (AlexanderYastrebov) [reply dataclients/kubernetes/kube.go]: Please resolve addresses comments

**Comment by AlexanderYastrebov** [comment]
I am wondering if we just should embed Ingress rule index into route id to ensure uniqueness?

**Comment by szuecs** [comment]
Another idea would be to remove the uniqueness for routeid or drop it completely.
Why do we need an id for a route?

**Comment by AlexanderYastrebov** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:
